### ISSUE TITLE
Running out of memory

### ISSUE THREAD

**Issue by ReedOnly** [vtsuperdarn/davitpy#299]
Hi,

I'm plotting a few hundred superDARN LoS plots at the time, and I'm running into the problem of memory overflow. After 70-80 plots Python takes up ca 8 GB memory, and if not shut down it crashes as it approaches 16 GB(total RAM of my computer).

I've been adding a few commands in the `pydarn.plotting.fan.plotFan` file to close the current figure, with no luck.

As of now I'm running:
```
myFig.clear()               
plot.clf()
plot.close(myFig)
pylab.close(myFig)
```

and none of them seem to remove my memory issue. Wondering if there might be the myMap Basemap object that keep filling up the memory? Any Ideas for how to completely clear the memory after each time time the `plotFan` function is called?

Regards,
Kristian Reed

**Comment by kkotyk** [comment]
This actually happens any time files are read if I remember correctly and you cannot properly process many files with DavitPy. Ashton can comment more, but there is a memory leak in the DMAP lib. I had tried to use DavitPy to process a lot of files and the work around I used involved using the Python multiprocessing module to spawn all DavitPy calls in it's own process so that the memory is force cleared when it exits. It's hacky, but it works.

Just a friendly reminder that this is one of the reasons I wrote a pure Python implementation of file reading/writing. It's a little slower than the C routines, but it's all Python, has only NumPy as a dependency, doesn't leak, and has more robust error handling.

**Comment by ReedOnly** [comment]
So you're saying the problems isn't really in the plots itself, but the `pydarn.sdio.radDataOpen` function loading the data? Guess there should be a way to shut down all the processes between every iteration of the loop where `pydarn.plotting.fan.plotFan` is called?

**Comment by kkotyk** [comment]
I'm not super familiar with the way DavitPy opens files, but I think that is correct. You will run into this problem with anything that involves batch processing files because any time a file is read, it leaks a significant amount of memory. You will definitely need a way to shut down the process involved with DavitPy calls. In my case I wrote a function to open files and produce RTI plots. I then use multiprocessing to spawn this function in it's own process and therefore all memory is freed when it exits. It's definitely not the way that should be handled, but the nice thing about it is that you can create lots of plots in parallel as multiprocessing will let you spawn a new process for every thread in your computer.

**Comment by asreimer** [comment]
Yes, @kkotyk is correct in saying there is a memory leak in reading file. I think it has something to do with pointers not getting freed properly, but I haven't figured it out just yet. I have ideas on how to fix things but haven't found the time to implement them.

The best solution in the mean time is do what @kkotyk suggested. Another alternative is to write a python script that you can call from the command line with `python some_script.py arg1 arg2 ...` and then call that script from within a shell script. The file reading and plotting for one file (which would be passed as an argument to `some_script.py`) would be handled performed for just that one file. Then the python interpreter would exit, freeing up the memory. This would prevent the leak from piling up over time.

**Comment by asreimer** [comment]
Duplicate of #191
